crates/fuel-core/src/query/balance.rs

```diff
@@ -36,22 +40,28 @@ impl ReadView {
         asset_id: AssetId,
         base_asset_id: AssetId,
     ) -> StorageResult<AddressBalance> {
-        let amount = AssetQuery::new(
-            &owner,
-            &AssetSpendTarget::new(asset_id, u64::MAX, u16::MAX),
-            &base_asset_id,
-            None,
-            self,
-        )
-        .coins()
-        .map(|res| res.map(|coins| coins.amount()))
-        .try_fold(0u64, |balance, amount| {
-            async move {
-                // Increase the balance
-                Ok(balance.saturating_add(amount))
-            }
-        })
-        .await?;
+        let amount = if self.balances_enabled {
+            debug!(%owner, %asset_id, "Querying balance with balances cache");
+            self.off_chain.balance(&owner, &asset_id, &base_asset_id)?
+        } else {
+            debug!(%owner, %asset_id, "Querying balance without balances cache");
+            AssetQuery::new(
+                &owner,
+                &AssetSpendTarget::new(asset_id, u64::MAX, u16::MAX),
```

Comment:
mostly a question form my side, what is an `AssetSpendTarget`? In particular, I'm trying to understand the second parameter in this call.

Reply:
`AssetSpendTarget` is a helper struct used to query for coins. My understanding is that it's main purpose is to make it easier to reply to "coins_to_spend()" requests, in which case the interpretation is as follows:
* `asset_id` - asset id :)
* `target` - the second parameter - how much value I want to get from this asset
* `max` - maximum number of coins representing this asset that I want to spend

For example: _"I need 1000 LCK but I don't want to use more than 3 coins"_ will be represented as `AssetSpendTarget::new("LCK", 1000, 3)`

Now, in the context of balances, we re-use this helper, but pass `u64::MAX, u16::MAX` to indicate that we don't want to impose limits on our query.